### PR TITLE
Add more checks to emitc::CallOp verifier

### PR DESCRIPTION
**Pull request by OliverScherf**
This adds the following checks to the `emitC::CallOp` verifier:
- args: arguments need a type
- template_args: arguments need a type, forbid float literals, forbid dense elements attributes

### CODE REVIEW

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 1]
```diff
@@ -50,6 +50,65 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+static LogicalResult verify(mlir::emitc::CallOp op) {
+  // callee must not be empty.
+  if (op.callee().empty()) {
+    return op.emitOpError("callee must not be empty");
+  }
+
+  auto argsAttr = op.args();
+  if (argsAttr.hasValue()) {
+    for (auto &arg : argsAttr.getValue()) {
+      // args elements of type index must be in range [0..operands.size).
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:66]:
```suggestion
      // Args with elements of type index must be in range [0..operands.size).
```

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 2]
```diff
@@ -50,6 +50,65 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+      if (auto iArg = arg.dyn_cast<IntegerAttr>()) {
+        if (iArg.getType().isIndex()) {
+          int64_t index = iArg.getInt();
+          if ((index < 0) ||
+              (index >= static_cast<int64_t>(op.getNumOperands()))) {
+            return op.emitOpError("index argument is out of range");
+          }
+        }
+      }
+      // args elements of type ArrayAttr must have a type.
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:76]:
```suggestion
     // Args with elements of type ArrayAttr must have a type.
```

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 3]
```diff
@@ -50,6 +50,65 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+  }
+
+  auto templateArgsAttr = op.template_args();
+  if (templateArgsAttr.hasValue()) {
+    for (auto &tArg : templateArgsAttr.getValue()) {
+      // C++ forbids float literals as template arguments.
+      if (auto iArg = tArg.dyn_cast<FloatAttr>()) {
+        return op.emitOpError("float literal as template argument is invalid");
+      } else if (auto aArg = tArg.dyn_cast<ArrayAttr>()) {
+        // template_args elements of type ArrayAttr must have a type.
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:92]: I think if rephrased, those sentences can start with a capital letter.

**Review by marbre — CHANGES_REQUESTED** [review]
Will take a closer look tomorrow.

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp (hunk 3 above)
Comment (OliverScherf) [reply lib/Dialect/EmitC/IR/EmitCDialect.cpp:92]: I rephrased it, please take a look at it

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 4]
```diff
@@ -50,6 +50,60 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+  }
+
+  auto templateArgsAttr = op.template_args();
+  if (templateArgsAttr.hasValue()) {
+    for (auto &tArg : templateArgsAttr.getValue()) {
+      // C++ forbids float literals as template arguments.
+      if (auto iArg = tArg.dyn_cast<FloatAttr>()) {
+        return op.emitOpError("float literal as template argument is invalid");
+      }
+      // Template args with elements of type array are not allowed
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:92]:
```suggestion
      // Template args with elements of type ArrayAttr are not allowed
```

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 5]
```diff
@@ -50,6 +50,60 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+static LogicalResult verify(mlir::emitc::CallOp op) {
+  // Callee must not be empty.
+  if (op.callee().empty()) {
+    return op.emitOpError("callee must not be empty");
+  }
+
+  auto argsAttr = op.args();
+  if (argsAttr.hasValue()) {
+    for (auto &arg : argsAttr.getValue()) {
+      // Args with elements of type index must be in range [0..operands.size).
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:66]: Maybe this comment should be moved down to the inner if-clause?

**Review by marbre — CHANGES_REQUESTED** [review]
Only minor nits.

File: lib/Dialect/EmitC/IR/EmitCDialect.cpp [hunk 6]
```diff
@@ -50,6 +50,61 @@ void emitc::buildTerminatedBody(OpBuilder &builder, Location loc) {
...
+  }
+
+  auto templateArgsAttr = op.template_args();
+  if (templateArgsAttr.hasValue()) {
+    for (auto &tArg : templateArgsAttr.getValue()) {
+      // C++ forbids float literals as template arguments.
+      if (auto iArg = tArg.dyn_cast<FloatAttr>()) {
+        return op.emitOpError("float literal as template argument is invalid");
+      }
+      // Template args with elements of type ArrayAttr are not allowed
```
Comment (marbre) [inline lib/Dialect/EmitC/IR/EmitCDialect.cpp:93]:
```suggestion
      // Template args with elements of type ArrayAttr are not allowed.
```

One more, didn't noticed it before :see_no_evil:

**Review by marbre — COMMENTED** [review]
Looks good to me, but let's wait for Simon's approval.

**Review by marbre — COMMENTED** [review]
(Needed to re-review to get rid of the changes requested flag... which obviously does not work. Will approve after Simon did.)

**Review by simon-camp — APPROVED** [review]
LGTM :)

**Review by marbre — APPROVED** [review]

**Comment by marbre** [comment]
I messed up during approving changes and re-requested reviews, but @simon-camp as well as I actually approved this.

**Pull request merged**